ceph: schedule mons across failure domains

noahdesu opened the pull request:
**Description of your changes:**
ceph: schedule mons across failure domains

**Which issue is resolved by this Pull Request:**
Resolves #2603 

this patch introduces failure-domain aware scheduling for monitor pods.
previously the only policy was to avoid scheduling monitors on the same
nodes. the new policy tries to spread monitors across zones and nodes.

   NOTE: this patch only enforces the well-known k8s "zone"
   failure-domain label (i.e. it does not look for region labels). this
   decision is made for two reasons. first, eliminating explicit
   scheduling is an important goal and will be done as part of upcoming
   work. this will enable all supported failure-domain labels to be
   handled. second, the "zone" domain is likely more appropriate for a
   larger portion of users (as opposed to the geographical distinction
   made with region), and supporting multiple failure domain types would
   add significant complexity for a temporary resolution to failure
   domain scheduling.

at a high-level this patch introduces a data structure `NodeUsage` that
is a pairing of a v1.Node and metadata relevant to scheduling: number of
monitor pods on the node, and its status w.r.t. to being scheduable for
new monitor pods.

each scheduling event computes a unified data structure that organizes
for each node a `NodeUsage` structure into per-failure-zone groups.
subsequent algorithms rely soley on this unified structure to make
scheduling decisions.

there are two entry points to the changes made:

1) during orchestration mon:Cluster:assignMons is invoked to make a
place new monitor pods onto nodes. the core scheduling logic is now
self-contained in mon:scheduleMonitor which implements node and zone
aware scheduling policies.

2) periodic health checks resolve any conflicts that may occur due to
changes to the cluster (e.g. new nodes, configuration changes, etc...).
the two existing checks: mons on invalid nodes, and overloaded nodes are
both handled.

fixes: #2603

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

noahdesu commented:
@BlaineEXE this should fix the mon rescheduling on `kubectl cordon`

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -380,37 +380,129 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
-			logger.Debugf("mon %s already assigned to a node, no need to assign", m.DaemonName)
+	// ensure all monitors have a node assignment. note that this isn't
+	// necessarily optimal: it does not try to move existing monitors which is
+	// handled by the period monitor health checks.
```
typo: `period[ic]`?

noahdesu replied:
yes

noahdesu commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -366,3 +312,127 @@ func removeMonitorFromQuorum(context *clusterd.Context, clusterName, name string
+	//  - does an empty node exist
+	emptyZone := false
+	emptyNode := false
+	for zi := range nodeZones {
```
Makes sense here, but elsewhere in this patch the index iteration is because the range-over-value construct creates value copies but we are returning pointers into the data structure. iterating over value references would be awesome

BlaineEXE commented:
For others who might not know, you can get this pull request directly from `rook/rook` via `git fetch upstream pull/3355/head:upstream-noah-mon-sched2` assuming `upstream` is the remote name for `rook/rook` on your machine.

Manual testing notes:
( this is still a WIP, so expect edits here )


- [ ] 3 node setup, install Ceph, cordon one of the nodes; expect that the mons don't change
- [ ] more to come...

noahdesu commented:
@BlaineEXE a handy trick I learned recently. add to ~/.gitconfig

```
[alias]
  pr = !"pr() { git fetch origin pull/$1/head:pr-$1; }; pr"
```

noahdesu commented on cluster/examples/kubernetes/ceph/cluster.yaml:
```diff
@@ -31,7 +31,7 @@ spec:
   # set the amount of mons to be started
   mon:
     count: 3
-    allowMultiplePerNode: false
```
yep. the issue is that the previous version of mon scheduling behaved as if this flag was true on _cluster creation_. so you could use this crd to start 3 mons in minikube, for example. i suppose non-production should use a different crd or change this manually?

noahdesu commented:
@BlaineEXE thanks for the thorough testing! i updated the pr the fix the failure issues you saw, and then provided output for my testing scenarios that corresponded to the check boxes above which I think were test cases that are important but hadn't been yet tested.

### 3 nodes

> FAIL: no nodes have AZ labels, install Ceph w/ 3 mons, cordon one of the nodes; expect that the mons don't change
> * mons don't change, but the operator eventually crashes trying to create mon d (4th mon) as a replacement for mon on cordoned node

This is fixed in ada09d59141d9efe92784a326ad8a96c2dbc8524 (will squash). There was no logic change; the check for NoSchedule was not robust enough.

### 5 nodes

> FAIL: no nodes have AZ labels, install Ceph w/ 3 mons, cordon one of the mon nodes; expect that the > * mons don't change
operator creataes mon d (4th mon), and then removes mon on cordoned node

Same fix as previous.

> FAIL uncordon cordoned node from test above; expect mons to stay where they are
> mons indeed do not move, so the behavior is correct, but the operator reports this log line, op-cluster: Updated node k8s-worker-0 is not valid and could not get added to cluster in namespace rook-ceph. which seems wrong to me: the node is now valid, and it can be added to the cluster; however, no mon is being scheduled

This error message is originating in the cluster controller from the code added to allow dynamic k8s node additions (https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/controller.go#L428). I think the error message is misleading---in any case the error message occurs when the k8s node update watcher skips orchestration due to the update to a _invalid_ node. I think this is a separate topic as long as it isn't affecting the test cases.

> FAIL changing the AZs on nodes during runtime should cause Rook to consider rebalancing the mons
either node update events are not being triggered by k8s, or the events are being discarded by the node change listener

Make sure to wait about 45 seconds for the monitor health thread to trigger a check. Adding an explicit trigger for k8s node update would be a good enhancement.

I confirmed this is working with the following test where there are 3 unlabeled mon nodes, then after adding an AZ label to a non-AZ labeled node a node is migrated to prefer the specific AZ information.

```
[nwatkins@smash rook]$ kubectl -n rook-ceph get pods -owide | grep mon
rook-ceph-mon-a-84b46586cd-7n825      1/1     ...        worker0   <none>           <none>
rook-ceph-mon-b-7c649c7677-8r6tz      1/1     ...        worker1   <none>           <none>
rook-ceph-mon-c-86bfb57f4d-z64gd      1/1     ...        worker2   <none>           <none>

[nwatkins@smash rook]$ kubectl -n rook-ceph label nodes worker3 failure-domain.beta.kubernetes.io/zone=z1
node/worker3 labeled

Less than a minute later

[nwatkins@smash rook]$ kubectl -n rook-ceph get pods -owide | grep mon
rook-ceph-mon-a-84b46586cd-7n825      0/1     Terminating   ...   worker0   <none>           <none>
rook-ceph-mon-b-7c649c7677-8r6tz      1/1     Running   ...    worker1   <none>           <none>
rook-ceph-mon-c-86bfb57f4d-z64gd      1/1     Running   ...    worker2   <none>           <none>
rook-ceph-mon-d-5d6d5c7599-t79hh      1/1     Running   ...   worker3   <none>           <none>

The final state

[nwatkins@smash rook]$ kubectl -n rook-ceph get pods -owide | grep mon
rook-ceph-mon-b-7c649c7677-8r6tz      1/1     ...    worker1   <none>           <none>
rook-ceph-mon-c-86bfb57f4d-z64gd      1/1     ...    worker2   <none>           <none>
rook-ceph-mon-d-5d6d5c7599-t79hh      1/1     ...    worker3   <none>           <none>
```

> mons on AZ-less nodes should reschedule to nodes which get AZ afterwards
should work when AZs can be changed at runtime?

This works at runtime (see previous example). As far as the actual policy of preferring AZ-labeled nodes the reasoning is that if AZ information is available, it should be used.

> all nodes have different AZ set, install Ceph w/ 3 mons, change one mon node's AZ to match a different mon node's AZ; expect a mon to be rescheduled and expect that new mon is created before the old one is destroyed (existing mon fault tolerance should not be reduced)
should work when AZs can be changed at runtime?

```
here is the current labeling

[nwatkins@smash rook]$ kubectl get nodes -o go-template=$'{{range .items }}{{index .metadata.name}} {{index .metadata.labels "failure-domain.beta.kubernetes.io/zone"}}\n{{end}}\n'
master m
worker0 z0
worker1 z1
worker2 z2
worker3 z3

[nwatkins@smash rook]$ kubectl -n rook-ceph get pods -owide | grep mon
rook-ceph-mon-c-86bfb57f4d-z64gd      1/1     Running     ...    worker2   <none>           <none>
rook-ceph-mon-d-5d6d5c7599-t79hh      1/1     Running     ...    worker3   <none>           <none>
rook-ceph-mon-e-74cd546454-rjkq9      1/1     Running     ...    worker0   <none>           <none>

now change worker2 to be in zone z0. the mon-e and mon-c will be in zone z0.

[nwatkins@smash rook]$ kubectl get nodes -o go-template=$'{{range .items }}{{index .metadata.name}} {{index .metadata.labels "failure-domain.beta.kubernetes.io/zone"}}\n{{end}}\n'
master m
worker0 z0
worker1 z1
worker2 z0
worker3 z3

that would cause z0 to be overloaded relative to an empty zone (e.g. z1) a new monitor will be created to replace a node in z0. the resulting placement after mon health runs correctly rebalances:

nwatkins@smash rook]$ kubectl -n rook-ceph get pods -owide | grep mon
rook-ceph-mon-c-86bfb57f4d-z64gd      1/1     ...    worker2   <none>           <none>
rook-ceph-mon-d-5d6d5c7599-t79hh      1/1     ...    worker3   <none>           <none>
rook-ceph-mon-f-74495754b5-tgfp4      1/1     ...   worker1   <none>           <none>
```

> install v1.0 w/ 3 mons, set three mons up in the same AZ, have other AZs on remaining nodes, upgrade to this PR master; expect that 2 mons move to other AZs and expect that the new mons are created before the old ones are destroyed (existing mon fault tolerance should not be reduced)

```
[nwatkins@smash rook]$ kubectl -n rook-ceph logs rook-ceph-operator-58fdc4c59d-gxbfp | head -1
2019-07-09 21:54:35.974787 I | rookcmd: starting Rook v1.0.3 with arguments '/usr/local/bin/rook ceph operator'

zone labels running 1.0.3

[nwatkins@smash rook]$ kubectl get nodes -o go-template=$'{{range .items }}{{index .metadata.name}} {{index .metadata.labels "failure-domain.beta.kubernetes.io/zone"}}\n{{end}}\n'
master <no value>
worker0 z0
worker1 z0
worker2 z0
worker3 z3

and the placement

[nwatkins@smash rook]$ kubectl -n rook-ceph get pods -owide | grep mon
rook-ceph-mon-a-85ff6f65b-mh7tg       1/1     ...  worker0   <none>           <none>
rook-ceph-mon-b-66944794fc-7k4rt      1/1     ...   worker1   <none>           <none>
rook-ceph-mon-c-5b4b598fb6-vkggz      1/1     ...  worker2   <none>           <none>

switch to local build

#> kubectl -n rook-ceph set image deploy/rook-ceph-operator rook-ceph-operator=rook/ceph:master

new mon starts before old is removed:

[nwatkins@smash rook]$ kubectl -n rook-ceph get pods -owide | grep mon
rook-ceph-mon-a-85ff6f65b-mh7tg       1/1     ...   worker0   <none>           <none>
rook-ceph-mon-b-66944794fc-7k4rt      1/1     ...  worker1   <none>           <none>
rook-ceph-mon-c-5b4b598fb6-vkggz      1/1     ...   worker2   <none>           <none>
rook-ceph-mon-d-dbd7d94df-tcwcl       1/1     ...  worker3   <none>           <none>

and now two mons are running zone z0 and one was move to z3.

[nwatkins@smash rook]$ kubectl -n rook-ceph get pods -owide | grep mon
rook-ceph-mon-b-66944794fc-7k4rt      1/1     ...   worker1   <none>           <none>
rook-ceph-mon-c-5b4b598fb6-vkggz      1/1     ...  worker2   <none>           <none>
rook-ceph-mon-d-dbd7d94df-tcwcl       1/1     ...   worker3   <none>           <none>
```

noahdesu commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -159,15 +159,9 @@ func (c *Cluster) checkHealth() error {
-	done, err := c.checkMonsOnValidNodes()
+	// find any mons that invalidate our placement policy, and if necessary,
+	// reschedule them to other nodes.
+	done, err := c.resolveInvalidMonitorPlacement(desiredMonCount)
```
correct. once the monitor is placed on a node, node cordon will not cause the monitor to be moved. this is my reading of the documentation. i only tested for `kubectl cordon node` not sure about other types of taints that we want to test.

noahdesu replied:
these scenarios are discussed more in the top-level discussion thread.

noahdesu commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -366,3 +312,127 @@ func removeMonitorFromQuorum(context *clusterd.Context, clusterName, name string
+	for name, nodeInfo := range c.mapping.Node {
+		if nodeInfo.Name == nodeChoice.Node.Name {
+			logger.Infof("rebalance: rescheduling mon %s from node %s", name, nodeInfo.Name)
+			c.failMon(len(c.clusterInfo.Monitors), desiredMonCount, name)
```
the naming of `failMon` is misleading. the selected monitor will be failed only if the monitor count is larger than the target, otherwise it will try to fail over / move the monitor to a different node.

BlaineEXE replied:
The behavior I've observed is that the new mon is created and added to the cluster before the current one is removed. Does that seem sufficient @leseb ?

leseb replied:
Yes, it is. Thanks.

BlaineEXE commented:
This introduces a fair bit of logging with the two below statements. I wonder if these would be better as Debug and only Info when Rook is about to act on new information.

```
2019-07-10 20:22:20.261982 I | op-mon: rebalance: no mon placement violations or fixes available
2019-07-10 20:23:05.742905 I | op-mon: rebalance: desired mon count: 3, empty zone found: false, empty node found: true
2019-07-10 20:23:05.742931 I | op-mon: rebalance: no mon placement violations or fixes available
2019-07-10 20:23:51.289610 I | op-mon: rebalance: desired mon count: 3, empty zone found: false, empty node found: true
2019-07-10 20:23:51.289629 I | op-mon: rebalance: no mon placement violations or fixes available
2019-07-10 20:24:36.756097 I | op-mon: rebalance: desired mon count: 3, empty zone found: false, empty node found: true
2019-07-10 20:24:36.759856 I | op-mon: rebalance: no mon placement violations or fixes available
2019-07-10 20:25:22.266294 I | op-mon: rebalance: desired mon count: 3, empty zone found: false, empty node found: true
2019-07-10 20:25:22.266443 I | op-mon: rebalance: no mon placement violations or fixes available
2019-07-10 20:26:07.755884 I | op-mon: rebalance: desired mon count: 3, empty zone found: false, empty node found: true
2019-07-10 20:26:07.756134 I | op-mon: rebalance: no mon placement violations or fixes available
```

noahdesu commented:
@BlaineEXE ahh, that's in the health check. yeh you're right i'll quiet that down.

BlaineEXE commented on pkg/operator/k8sutil/node.go:
```diff
@@ -34,13 +34,9 @@ var validTopologyLabelKeys = []string{
-	}
-
+// ValidNodeNoSched returns true if the node (1) meets Rook's placement terms,
+// and (2) is ready. False otherwise.
```
I think a note here that a node *is* allowed to have a `NoSchedule` taint would be useful for the future

noahdesu replied:
fixed

BlaineEXE commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -366,3 +314,133 @@ func removeMonitorFromQuorum(context *clusterd.Context, clusterName, name string
+		// this case occurs if the zone has more than 1 monitor, and some other
+		// zone exists without any monitors.
+		if nodeChoice == nil && zoneMonCount > 1 && emptyZone {
+			for ni := range nodeZones[zi] {
```
This for loop is not captured in unit test coverage.

noahdesu replied:
nice catch. does the the normal `make check` give me coverage? I hadn't seen this but maybe didn't know I could look for it.

BlaineEXE replied:
I found it by turning coverage highlighting on in VSCode. I'm really not sure what it uses under the covers.

noahdesu replied:
oh, thanks. yeh i found this `go test -coverprofile=coverage.out; go tool cover -func=coverage.out`  opens a browser with the coverage results.

noahdesu replied:
fixed

BlaineEXE commented on pkg/operator/ceph/cluster/mon/node.go:
```diff
@@ -21,85 +21,17 @@ import (
-		nodesInUse.Add(name)
-	}
-	return nodesInUse, nil
+type NodeUsage struct {
```
Add godoc comment please.

noahdesu replied:
fixed

BlaineEXE commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -366,3 +314,133 @@ func removeMonitorFromQuorum(context *clusterd.Context, clusterName, name string
+			logger.Infof("rebalance: rescheduling mon %s from node %s", name, nodeInfo.Name)
+			c.failMon(len(c.clusterInfo.Monitors), desiredMonCount, name)
+			return true, nil
+		} else {
```
FWIW my linter is complaining that `if block ends with a return statement, so drop this else and outdent its block`

noahdesu replied:
fixed

BlaineEXE commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -366,3 +314,133 @@ func removeMonitorFromQuorum(context *clusterd.Context, clusterName, name string
+
+	for zi := range nodeZones {
+		zoneMonCount := 0
+		var nodeChoice *NodeUsage = nil
```
FWIW My linter complains `if block ends with a return statement, so drop this else and outdent its block`

noahdesu replied:
fixed

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -380,37 +380,129 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+// they are in distinct zones.
+func scheduleMonitor(mon *monConfig, nodeZones [][]NodeUsage) *NodeUsage {
+	// the node choice for this monitor
+	var nodeChoice *NodeUsage = nil
```
FWIW My linter complains if block ends with a return statement, so drop this else and outdent its block

noahdesu replied:
fixed

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -380,37 +380,129 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+		zoneMonCount := 0
+
+		// preferential node choice from this zone for a new mon
+		var zoneNodeChoice *NodeUsage = nil
```
FWIW My linter complains if block ends with a return statement, so drop this else and outdent its block

noahdesu replied:
fixed

noahdesu commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -366,3 +312,127 @@ func removeMonitorFromQuorum(context *clusterd.Context, clusterName, name string
+			c.failMon(len(c.clusterInfo.Monitors), desiredMonCount, name)
+			return true, nil
+		} else {
+			logger.Warningf("rebalance: no mon pod found on node %s", nodeChoice.Node.Name)
```
yep, fixed.

noahdesu commented:
> OK??? : install w/ 3 mons, force shutdown a node w/ mon scheduled; expec the mon to be rescheduled within 5-6 minutes
> was rescheduled within the minute it seemed ... is this the right behavior?

@BlaineEXE did you just power down the node and nothing else?

BlaineEXE commented:
> > OK??? : install w/ 3 mons, force shutdown a node w/ mon scheduled; expec the mon to be rescheduled within 5-6 minutes
> > was rescheduled within the minute it seemed ... is this the right behavior?
> 
> @BlaineEXE did you just power down the node and nothing else?

Yes. I just force shutdown the VM. This might simulate a node failure due to power loss or sudden network partition.

noahdesu commented:
> OK??? : install w/ 3 mons, force shutdown a node w/ mon scheduled; expec the mon to be rescheduled within 5-6 minutes
was rescheduled within the minute it seemed ... is this the right behavior?

@BlaineEXE FWIW I tried this on v1.0 and I get the same behavior.

```
2019-07-11 21:06:25.650297 W | op-mon: mon b not found in quorum, waiting for timeout before failover
2019-07-11 21:07:11.005983 W | op-mon: mon b not found in quorum, waiting for timeout before failover
2019-07-11 21:07:11.009584 W | op-mon: node worker1 isn't valid anymore, failover mon b
2019-07-11 21:07:11.009601 I | op-mon: Failing over monitor b
2019-07-11 21:07:11.009617 I | op-mon: starting new mon: &{ResourceName:rook-ceph-mon-d DaemonName:d PublicIP: Port:6789 DataPathMap:0xc000af8550}                                                                                                                                       
2019-07-11 21:07:11.032779 I | op-mon: mon d endpoint are [v2:10.99.133.133:3300,v1:10.99.133.133:6789]
2019-07-11 21:07:11.052195 I | op-mon: Found 0 running nodes without mons
2019-07-11 21:07:11.052244 I | op-mon: All nodes are running mons. Adding all 3 nodes to the availability.
2019-07-11 21:07:11.057195 I | op-mon: mons created: 1
2019-07-11 21:07:11.057216 I | op-mon: waiting for mon quorum with [d]
2019-07-11 21:07:11.065393 I | op-mon: mon d is not yet running
2019-07-11 21:07:11.065415 I | op-mon: mons running: []
2019-07-11 21:07:16.076452 I | op-mon: mons running: [d]
```

noahdesu commented:
squashed and rebased. i believe that i addressed all the feedback. could use testing, especially from anyone that hasn't yet tested.
